server/src/main/java/org/elasticsearch/indices/recovery/MultiFileTransfer.java

```diff
@@ -0,0 +1,171 @@
+                if (done.compareAndSet(false, true)) {
+                    listener.onResponse(r);
+                }
+            },
+            e -> {
+                if (done.compareAndSet(false, true)) {
+                    listener.onFailure(e);
+                }
+            });
+        this.processor = new AsyncIOProcessor<>(logger, maxConcurrentFileChunks * 2, threadContext) {
+            @Override
+            protected void write(List<Tuple<FileChunkResponseItem<Response>, Consumer<Exception>>> items) {
+                handleItems(items);
+            }
+        };
+        this.remainingFiles = files.iterator();
+    }
+
+    public final void start() {
+        // put an dummy item to start the processor
+        processor.put(new FileChunkResponseItem<>(SequenceNumbers.UNASSIGNED_PRIMARY_TERM, null, null, null), e -> {});
+    }
+
+    private void handleItems(List<Tuple<FileChunkResponseItem<Response>, Consumer<Exception>>> items) {
+        if (done.get()) {
+            return;
+        }
+        try {
+            for (Tuple<FileChunkResponseItem<Response>, Consumer<Exception>> item : items) {
+                final FileChunkResponseItem<Response> resp = item.v1();
+                if (resp.requestSeqId == SequenceNumbers.UNASSIGNED_SEQ_NO) {
+                    continue; // not an actual item
+                }
+                requestSeqIdTracker.markSeqNoAsProcessed(resp.requestSeqId);
+                if (resp.failure != null) {
+                    handleError(resp.md, resp.failure);
+                    throw resp.failure;
+                }
+            }
+            while (requestSeqIdTracker.getMaxSeqNo() - requestSeqIdTracker.getProcessedCheckpoint() < maxConcurrentFileChunks) {
+                if (currentFile == null) {
+                    if (remainingFiles.hasNext()) {
+                        currentFile = remainingFiles.next();
+                    } else {
+                        if (requestSeqIdTracker.getProcessedCheckpoint() == requestSeqIdTracker.getMaxSeqNo()) {
+                            listener.onResponse(null);
+                        }
+                        return;
+                    }
+                }
+                final Request request;
+                try {
+                    request = nextChunkRequest(currentFile, fileOffset);
+                } catch (Exception e) {
+                    handleError(currentFile, e);
+                    throw e;
+                }
+                final long requestSeqId = requestSeqIdTracker.generateSeqNo();
+                final StoreFileMetaData md = this.currentFile;
+                sendChunkRequest(request, ActionListener.wrap(
+                    r -> processor.put(new FileChunkResponseItem<>(requestSeqId, md, r, null), ignored -> {}),
+                    e -> processor.put(new FileChunkResponseItem<>(requestSeqId, md, null, e), ignored -> {})
+                ));
+                fileOffset += request.sizeInBytes();
```

Comment:
We are checking for lastChunk in two places now, would it be an idea to expose `lastChunk()` on `ChunkRequest`?

Also, I think it would be nicer track the offset in the subclass, it seems like a shared responsibility now.

Reply:
++